### PR TITLE
Task/sapnamysore/tlt 1805/publish courses

### PR DESCRIPTION
**Pull request by sapnamysore**
This PR contains changes for the [tlt-1805](https://jira.huit.harvard.edu/browse/TLT-1805) : a new tool to Bulk publish Canvas sites.  It includes the front end, back end and bulk utilities. The unit tests and Jasmine tests are being worked on in parallel, and shouldn't hold up this PR . 

It depends on this PR :https://github.com/Harvard-University-iCommons/django-icommons-common/pull/169 
[Note : once the above PR is merged, it's reference in  base.text will need to be fixed]

### CODE REVIEW

File: canvas_account_admin_tools/static/canvas_account_admin_tools/js/services/ATRAPIService.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
+(function() {
+  var module = angular.module('ATRAPIModule', []);
```
Comment (sapnamysore) [inline canvas_account_admin_tools/static/canvas_account_admin_tools/js/services/ATRAPIService.js]: Nitpick:  I had mentioned this in the first pass :-). 
https://github.com/Harvard-University-iCommons/canvas_account_admin_tools/pull/184/files/28b98b54157fc9ee590cb78a356b07a3b0c10f54#diff-f7d9736aa0774b6bc2ae795dc8809e20R2

File: publish_courses/tests/test_async_operations.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Test TODOs
+# process status, timestamps updated when activated
```
Comment (elliottyates) [inline publish_courses/tests/test_async_operations.py]: (this was fixed by adding two more possible tests just below)

File: canvas_account_admin_tools/static/canvas_account_admin_tools/js/services/ATRAPIService.js (hunk 1 above)
Comment (sapnamysore) [inline canvas_account_admin_tools/static/canvas_account_admin_tools/js/services/ATRAPIService.js]: Maybe ActRestAPIModule (for Academic TechRestAPI) :-). But as discussed offline, this is a very low priority.

File: publish_courses/api.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            }
+            summary_counts_by_state = {'total': len(total_courses)}
+            for k, v in state_map.items():
+                summary_counts_by_state[k] = len(
+                    [c for c in total_courses if c['workflow_state'] == v])
+        except Exception as e:
+            logger.exception(
+                "Failed to get published courses summary for term_id={} and "
+                "account_id={}".format(self.term_id, self.account_id))
+            raise CanvasAPIError('There was a problem counting courses. '
```
Comment (sapnamysore) [inline publish_courses/api.py]: Agree, we could re-raise the exception with a new message

File: publish_courses/migrations/0001_initial.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+LTI_PERMISSIONS_DATA = [
+    ('publish_courses', '*', 'AccountAdmin', True),
```
Comment (sapnamysore) [inline publish_courses/migrations/0001_initial.py]: Hmm - I think we had 3 variations in the past and now we are down to 2? :-)

Comment (elliottyates) [inline publish_courses/migrations/0001_initial.py]: I think we no longer need the `AccountAdmin` (no space) version.

Comment (elliottyates) [inline publish_courses/migrations/0001_initial.py]: I checked canvas.dev and I'm showing up as an `Account Admin` -- maybe another person who's not Canadian can double-check that and we can remove this line (7).

**Comment by sapnamysore** [comment]
Merging the PR

File: publish_courses/api.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            }
+            summary_counts_by_state = {'total': len(total_courses)}
+            for k, v in state_map.items():
+                summary_counts_by_state[k] = len(
+                    [c for c in total_courses if c['workflow_state'] == v])
+        except Exception as e:
+            logger.exception(
+                "Failed to get published courses summary for term_id={} and "
+                "account_id={}".format(self.term_id, self.account_id))
+            raise RuntimeError("There was a problem counting courses. ")
```
Comment (elliottyates) [inline publish_courses/api.py]: Ah, I realize now that we had made the `CanvasAPIError` as a custom error class that was backed by a DRF base error class (similar to the [LTIPermission exception](https://github.com/Harvard-University-iCommons/canvas_account_admin_tools/pull/193/files#diff-520d38b2355a5dd75c3c0e0a4e235dbdR36) above). We probably should stick with that in the future, as those errors allow configuring the HTTP status code and the error message, etc. Ok for now though.

**Pull request merged**